File: flutter/supy_scanner/ios/Classes/log/SupyLog.swift
import Foundation
import os.log

/// Thin tagged wrapper around `os.Logger` used by every native diagnostic
/// site inside the plugin. Mirror of the Dart `SupyLog` facade
/// (`lib/src/log/supy_log.dart`). Keep the level enum and tag convention in
/// sync.
///
/// Host apps that want to silence the library entirely flip `enabled` to
/// `false` (e.g. from a debug menu).
///
/// Never log barcode payloads, OCR text, or file URIs through this — see
/// `docs/SECURITY.md` §8.
@objc public final class SupyLog: NSObject {
    /// Toggle the entire native log stream.
    @objc public static var enabled: Bool = true

    private static let subsystem = "io.supy.scanner"
    private static let defaultLogger = Logger(subsystem: subsystem, category: "SupyScanner")
    private static var loggers: [String: Logger] = [:]
    private static let lock = NSLock()

    private static func logger(for tag: String) -> Logger {
        if tag.isEmpty {
            return defaultLogger
        }
        lock.lock()
        defer { lock.unlock() }
        if let cached = loggers[tag] {
            return cached
        }
        let fresh = Logger(subsystem: subsystem, category: tag)
        loggers[tag] = fresh
        return fresh
    }

    @objc public static func d(_ message: String, tag: String = "SupyScanner") {
        guard enabled else { return }
        logger(for: tag).debug("\(message, privacy: .public)")
    }

    @objc public static func i(_ message: String, tag: String = "SupyScanner") {
        guard enabled else { return }
        logger(for: tag).info("\(message, privacy: .public)")
    }

    @objc public static func w(_ message: String, tag: String = "SupyScanner") {
        guard enabled else { return }
        logger(for: tag).warning("\(message, privacy: .public)")
    }

    @objc public static func e(_ message: String, tag: String = "SupyScanner") {
        guard enabled else { return }
        logger(for: tag).error("\(message, privacy: .public)")
    }
}

File: flutter/supy_scanner/android/src/main/kotlin/io/supy/scanner/log/SupyLog.kt
package io.supy.scanner.log

import android.util.Log

/**
 * Thin tagged wrapper around [android.util.Log] used by every native
 * diagnostic site inside the plugin. Host apps that want to silence the
 * library entirely flip [enabled] to false (e.g. from a debug menu).
 *
 * Mirror of the Dart `SupyLog` facade (`lib/src/log/supy_log.dart`). Keep
 * the level enum and tag convention in sync.
 *
 * Never log barcode payloads, OCR text, or file URIs through this — see
 * `docs/SECURITY.md` §8.
 */
object SupyLog {
    /** Default tag if a caller doesn't supply one. */
    private const val DEFAULT_TAG = "SupyScanner"

    /** Toggle the entire native log stream. */
    @Volatile
    @JvmStatic
    var enabled: Boolean = true

    @JvmStatic
    fun d(tag: String = DEFAULT_TAG, message: String) {
        if (!enabled) return
        Log.d(tag, message)
    }

    @JvmStatic
    fun i(tag: String = DEFAULT_TAG, message: String) {
        if (!enabled) return
        Log.i(tag, message)
    }

    @JvmStatic
    fun w(tag: String = DEFAULT_TAG, message: String, throwable: Throwable? = null) {
        if (!enabled) return
        if (throwable != null) {
            Log.w(tag, message, throwable)
        } else {
            Log.w(tag, message)
        }
    }

    @JvmStatic
    fun e(tag: String = DEFAULT_TAG, message: String, throwable: Throwable? = null) {
        if (!enabled) return
        if (throwable != null) {
            Log.e(tag, message, throwable)
        } else {
            Log.e(tag, message)
        }
    }
}
